geopandas/base.py

```diff
@@ -502,9 +502,13 @@ def buffer(self, distance, resolution=16, **kwargs):
             Optional, the resolution of the buffer around each vertex.
         """
         if isinstance(distance, (np.ndarray, pd.Series)):
-            if distance.shape[0] != self.shape[0]:
+            if len(distance) != len(self.index):
                 raise ValueError("Length of distance sequence does not match "
                                  "length of the GeoSeries")
+            if isinstance(distance, pd.Series):
+                if not (distance.index == self.index).all():
```

Comment:
you can use `self.index.equals(distance.index)` here, that will short-cut if they are the same object (eg if `distance` is coming from a column in the same dataframe, both indexes are the same object)